staging/src/k8s.io/kubectl/pkg/cmd/describe/describe.go

```diff
@@ -174,16 +174,22 @@ func (o *DescribeOptions) Run() error {
 
 	errs := sets.NewString()
 	first := true
+	describerCache := newDescriberCache()
 	for _, info := range infos {
 		mapping := info.ResourceMapping()
-		describer, err := o.Describer(mapping)
-		if err != nil {
-			if errs.Has(err.Error()) {
+		describer := describerCache.get(mapping)
+		if describer == nil {
+			var err error
+			describer, err = o.Describer(mapping)
```

Comment:
Important note! User facing behavior change!

- In this proposed change, we would be using the same `kubernetes.Clientset` to describe each individual resource in `infos`.
- That means that rate limiting will take effect.
- When running on my machine with a `kind` cluster and 100 `ConfigMaps`'s, `kubectl describe cm 1>/dev/null 2>/dev/null` slows down from about `1.5s` to
- Although! We already see this rate limiting take effect with the `kubectl describe cm some-prefix-` behavior (i.e., `describe.DescribeMatchingResources`), so maybe rate limiting in both cases is more congruent behavior.
- Regardless, this seems like a user-facing change, so I wanted to point this out.
- I don't feel great about this slowdown, even with the benefit that this change brings.
- I tried increasing the `QPS`/`Burst` on the client, but that seemed arbitrary and janky.
- I am curious of other people's thoughts here.

Reply:
> * When running on my machine with a `kind` cluster and 100 `ConfigMaps`'s, `kubectl describe cm 1>/dev/null 2>/dev/null` slows down from about `1.5s` to `38s`. :(

What's really unfortunate is that we've already fetched the objects from the server using a list call... ideally, we would just describe the object we fetched directly rather than re-fetching it

Reply:
Yeah, agreed. I was just taking a peek at that. Since we also (optionally) print out events associated with the object that we are describing, I believe we will need to make some sort of API call. However, if we reused the `info.Object` that would get rid of 1 API call...

When I make a code change to pass the `info.Object` directly to the `ConfigMapDescriber` and run the same test as I did above with 100 `ConfigMap`'s, I get a runtime of about `18s`. Improvement. The need for the searching of `Event`'s is still adding an order of magnitude, though.

My gut says to spin this improvement out into another GH issue.

Reply:
Issue opened: https://github.com/kubernetes/kubernetes/issues/97852.